### PR TITLE
Add support for glob syntax in srcJar includes

### LINKED ISSUE #1795: Allow filtering by path for srcJar dependencies
We use proto schema published as jar, but we need just a subset of that. But `include` option of `sourcePath` closure only works for directory dependencies. It would be nice to allow it for jar dependencies too

### PR DESCRIPTION
**Pull request by benjamin-bader**
When specifying proto root sets in the Gradle extension, you have a few options but generally it boils down to identifying some sort of root location, and optionally some `include` filters to narrow down which .proto files are included.  Roots can come in a few types - paths, projects, and jars (either local or remote).  Paths have a few superpowers compared to jars here, in that you can use glob syntax to concisely specify what files to include.

Jars don't support this today.  This is because glob file matching basically comes for free with Gradle file collections, but those don't work with jars.

This PR adds glob support to jar proto roots, using the JDK's ZipFileProvider to crack open jars and walk through them, converting glob notation to concrete `Location` instances.

It's maybe not the prettiest code I've ever written, but it seems to get the job done.  It's a little awkwardly-placed, in the `WireTask` itself.  This is necessary because with remote jars, we have no guarantee during configuration time that the jar is available on the local filesystem.  We have to wait for all dependencies to have been resolved, and that isn't guaranteed until the task is running.

Fixes #1795

### CODE REVIEW

**Review by oldergod — COMMENTED** [review]
Will review that in details soon. Thanks

File: wire-library/wire-gradle-plugin/src/test/kotlin/com/squareup/wire/gradle/WirePluginTest.kt [hunk 1]
```diff
@@ -267,6 +267,21 @@ class WirePluginTest {
...
+  @Test
+  fun sourceJarRemoteWildcardIncludes() {
+    val fixtureRoot = File("src/test/projects/sourcejar-remote-wildcards")
+
+    val result = gradleRunner.runFixture(fixtureRoot) { build() }
+
+    assertThat(result.task(":generateProtos")).isNotNull
+    assertThat(result.output)
+        .contains("Writing com.squareup.dinosaurs.Dinosaur")
+        .contains("Writing com.squareup.geology.Period")
```
[the remark this answers is not in the document]
Comment (benjamin-bader) [reply wire-library/wire-gradle-plugin/src/test/kotlin/com/squareup/wire/gradle/WirePluginTest.kt:279]: Good question, I'll look in to that

[the remark this answers is not in the document]
Comment (benjamin-bader) [reply wire-library/wire-gradle-plugin/src/test/kotlin/com/squareup/wire/gradle/WirePluginTest.kt:279]: There's only the two files in that jar.  I'd like to have both single- and double-start includes in the test, so I guess I'll add a new jar that has something to exclude in it.

[the remark this answers is not in the document]
Comment (benjamin-bader) [reply wire-library/wire-gradle-plugin/src/test/kotlin/com/squareup/wire/gradle/WirePluginTest.kt:279]: Done

**Comment by oldergod** [comment]
I tweaked a few things. Will merge on green, thank you.

**Pull request merged**